src/box/relay.cc

```diff
@@ -813,6 +845,56 @@ relay_thread_on_stop(void *arg)
 	relay->tx.is_raft_enabled = false;
 }
 
+/** The next_sync call message. */
+struct relay_next_sync_msg {
+	/** Parent cbus message. */
+	struct cbus_call_msg base;
+	/** The queried relay. */
+	struct relay *relay;
+	/** Sync value returned from relay. */
+	uint64_t sync;
+};
+
+/** A callback to free the message once it returns to tx thread. */
+static int
+relay_next_sync_msg_free(struct cbus_call_msg *msg)
+{
+	free(msg);
+	return 0;
+}
+
+/** Relay side of the next_sync call. */
+static int
+relay_next_sync_f(struct cbus_call_msg *msg)
+{
+	struct relay_next_sync_msg *m = (struct relay_next_sync_msg *)msg;
+	m->sync = m->relay->sync + 1;
+	m->relay->need_new_heartbeat = true;
+	return 0;
+}
+
+int
+relay_next_sync(struct relay *relay, uint64_t *sync, double deadline)
+{
+	struct relay_next_sync_msg *msg =
+		(struct relay_next_sync_msg *)xmalloc(sizeof(*msg));
```

Comment:
If you want to mAkE sOmE cOoL oPtImIsAtIoNs to improve batching, then you could make this `relay_next_sync_msg` a part of `struct relay` instead of allocating it on the heap. And it would work like iproto kharon and applier ack - go back and forth on demand. As a bonus, it could then be reused by multiple linerizable txns requesting next sync at the same time. If a txn comes and sees the sync req is not yet sent, then it pushes the message. If txns come and see the req is already sent, then they set a flag in relay that the msg should be re-sent and wait for it being re-sent and re-delivered.

When sync returns to tx thread, the first txn takes it and leaves. The others together re-request the sync. When it returns, they also all take it and leave.

Current approach is fine too, although not so good for batching. My only worry is that if relay thread is busy for too long, then there can be quite a lot of timed out 'next_sync' calls which leave `relay_next_sync_msg` garbage allocated on the heap until they complete.

Can leave it as is if you want.

Reply:
Yes, it's a good idea, but I'd rather not do that now. As we've discussed in the RFC, linearizable transactions aren't supposed to be fast. I think it's better to concentrate on correctness rather than speed right now.